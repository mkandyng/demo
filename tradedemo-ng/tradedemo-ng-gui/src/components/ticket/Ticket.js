import React, {useState, useEffect} from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { validateAndPlaceOrder } from "../../common/orderbook"
import { getRandomInt, toggleOpacity, getDateString } from "../../common/utils";
import { placeOrder, updateOrder } from "../../store/orderbook/orderbookActions"
import TicketView from "./TicketView";

function Ticket(props) {
    const { instrument,
            placeOrder,
            updateOrder } = props;

    const [ticket, updateTicket] = useState( {
        symbol: undefined,
        orderId:0,
        quantity: 5,
        orderType: "Market",
        price: 0,
        expiryType: "Day",
        expiryDate: "",
        note: "",
        priceStyle: {opacity: "0.5"},
        expiryDateStyle: {opacity: "0.5"}
    })

    useEffect ( () => {
        const MAX_ORDER_COUNT = 10;
        if(instrument.symbol !== undefined && ticket.orderId < MAX_ORDER_COUNT) {
            let buySell = getRandomInt(0,1) === 0 ? "Buy":"Sell";
            if(validateAndPlaceOrder({ ticket: ticket,
                                       instrument: instrument,
                                       buySell: buySell,
                                       price: instrument.price,
                                       confirmOrder: false,
                                       placeOrder: placeOrder,
                                       updateOrder: updateOrder })) {
                updateTicket({...ticket, orderId: ticket.orderId+1});
            }
        }
    }, [instrument, ticket, placeOrder, updateOrder]);

    const eventHandler = {
        symbolChange: e => updateTicket({...ticket, price: instrument.price.toFixed(2)}),
        expiryDateChange: e => updateTicket({...ticket, expiryDate: e.target.value}),
        quantityChange: e => updateTicket({...ticket, price: e.target.value}),
        priceChange: e => updateTicket({...ticket, price: e.target.value}),
        noteChange: e => updateTicket({...ticket, note: e.target.value}),
        orderTypeChange: e => updateTicket({...ticket,
                  priceStyle: {opacity: toggleOpacity(e.target.value, "Market", true)},
                  orderType: e.target.value,
                  price: instrument.price.toFixed(2)}),
        expiryTypeChange: e => updateTicket({...ticket,
                  expiryDateStyle: {opacity: toggleOpacity(e.target.value, "GTD", false)},
                  expiryType: e.target.value,
                  expiryDate: e.target.value === "GTD"?getDateString(new Date(), "dateOnly"):""}),
        handleOnSubmit: e => {
                  e.preventDefault();
                  const buySell = e.target.innerText;
                  if(validateAndPlaceOrder({ticket:ticket,
                                            instrument:instrument,
                                            buySell:buySell,
                                            price: Number(ticket.price),
                                            confirmOrder: true,
                                            placeOrder: placeOrder,
                                            updateOrder: updateOrder})) {
                      updateTicket({...ticket, orderId: ticket.orderId+1});
                  }
        }
    };

    return (
      <TicketView ticket={ticket}
                  instrument={instrument}
                  eventHandler={eventHandler} />
    )
}

const mapStateToProps = state => ({ instrument: state.marketfeed.selected });

// Map Redux actions to component props
const mapDispatchToProps = dispatch =>
     bindActionCreators({
        placeOrder,
        updateOrder
     }, dispatch);

// The HOC
export default connect(
     mapStateToProps,
     mapDispatchToProps
)(Ticket);
