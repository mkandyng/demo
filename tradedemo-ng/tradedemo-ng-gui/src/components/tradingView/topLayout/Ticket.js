import React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { placeOrder } from "../../../store/actions/placeOrder";
import { updateOrder } from "../../../store/actions/updateOrder";
import { placeOrderAndGenerateTradeLifeCycle } from "../../../libs/orderbook.js"
import { getRandomInt, toggleOpacity, getDateString } from "../../../libs/utils";

class Ticket extends React.Component {
    constructor(props) {
    	super(props)
	this.state = {
		       symbol: undefined,
		       orderId:1,
		       quantity: 5,
		       orderType: "Market",
		       price: 0,
		       expiryType: "Day",
		       expiryDate: "",
		       note: "",
		       priceStyle: {opacity: "0.5"},
		       expiryDateStyle: {opacity: "0.5"}	
		     }
 	this.symbolChange = this.symbolChange.bind(this);
	this.orderTypeChange = this.orderTypeChange.bind(this);
	this.expiryTypeChange = this.expiryTypeChange.bind(this);
        this.expiryDateChange = this.expiryDateChange.bind(this);
        this.quantityChange = this.quantityChange.bind(this)
        this.priceChange = this.priceChange.bind(this);
	this.noteChange = this.noteChange.bind(this);
        this.handleOnSubmit = this.handleOnSubmit.bind(this);
	this.validateAndPlaceOrder = this.validateAndPlaceOrder.bind(this);
    }

    componentDidMount() {
	const placeOrder = instrument => {
	    const placeInterval = setInterval(() => {
	        clearInterval(placeInterval);
		let buySell = getRandomInt(0,1) === 0 ? "Buy":"Sell";
		this.validateAndPlaceOrder(buySell, instrument, instrument.price, false);
	    }, getRandomInt(500,50000));
	};

	const interval = setInterval(() => {
	   const maxInstrumentsForDemo = 30;
	   if(this.props.marketfeedInstruments.length > 0) {
		clearInterval(interval);
		for(let i=0; i<maxInstrumentsForDemo; i++) {
		    this.props.marketfeedInstruments.forEach(instrument => {
			placeOrder(instrument);	
		    });
		}
	   }
	}, 300);
    }

    handleOnSubmit(e) {
	e.preventDefault();
   	const buySell = e.target.innerText;
	const instrument = this.props.marketfeedInstruments[this.props.selectedMarketfeedIndex];
	this.validateAndPlaceOrder(buySell, instrument, Number(this.state.price), true);
    }

    validateAndPlaceOrder(buySell, instrument, price, confirmOrder) {
        const validateOrder = () => {
	    if(this.state.quantity === 0) {
	        alert("Quantity must be greater than 0");
		return false;
	    }
            if(this.state.expiryType === "GTD") {
		var today = new Date();
		today.setHours(1,0,0,0);
		var inputDate = new Date(this.state.expiryDate);
		if(inputDate < today) {
		    alert("GTD date must be greater or equals to today [" + getDateString(today, "dateOnly") + "]");
		    return false;
		}
	    }
	    if(this.state.orderType !== "Market") {
	        const instrument = this.props.marketfeedInstruments[this.props.selectedMarketfeedIndex];
		if(Math.abs(this.state.price - instrument.price) > (instrument.price * 0.05)) {
		    alert("Price " + this.state.price + " is outside 5% of midPrice [" + instrument.price.toFixed(2) + "]");
		    return false;
		} 
	    }
	    return true;
	}
	if(validateOrder()) {
	    const priceInfo = this.state.orderType === "Market" ? "Market price": this.state.orderType + " (" + this.state.price + ")"; 
	    let confirmedPlaceOrder = true;
	    if(confirmOrder) {
	        //eslint-disable-next-line
		confirmedPlaceOrder = confirm(buySell + " " + this.state.quantity + " " + instrument.name + " @ " + priceInfo);
	    }
	    
	    if(confirmedPlaceOrder) {
	    	const padDigits = function(number, digits) {
    		    return Array(Math.max(digits - String(number).length + 1, 0)).join(0) + number;
	    	}
		placeOrderAndGenerateTradeLifeCycle(this.props, {
				    	orderRef:"XA"+ padDigits(this.state.orderId, 8),
				    	symbol:instrument.symbol,
					instrument: instrument.name,
					ccy: instrument.currency,
					buySell: buySell,
					expiryType: this.state.expiryType,
					expiryDate: this.state.expiryDate,
				        orderType: this.state.orderType,
					price: this.state.orderType === "Market" ? "": price,
					quantity: this.state.quantity,
					note: this.state.note,
					executed: 0,
					midPrice: instrument.price   
				    });
	        this.setState({orderId: this.state.orderId + 1});
	   }
	}
    }
   
    symbolChange(e) {
	this.setState( {
	    price: this.props.marketfeedInstruments[this.props.selectedMarketfeedIndex].price.toFixed(2)
	});
    }

    orderTypeChange(e) {
	this.setState({
	    priceStyle: {opacity: toggleOpacity(e.target.value, "Market", true)},
	    orderType: e.target.value,
	    price: this.props.marketfeedInstruments[this.props.selectedMarketfeedIndex].price.toFixed(2)
	});
    }

    expiryTypeChange(e) {
	this.setState({
	    expiryDateStyle: {opacity: toggleOpacity(e.target.value, "GTD", false)},
	    expiryType: e.target.value,
	    expiryDate: e.target.value === "GTD"?getDateString(new Date(), "dateOnly"):""
	});
    }

    expiryDateChange(e) {
	this.setState({
	    expiryDate: e.target.value
	});
    }
    
    quantityChange(e) {
	this.setState({
	    quantity: e.target.value
	});
    }
    
    priceChange(e) {
	this.setState({
            price: e.target.value
        });
    }

    noteChange(e) {
	this.setState({
            note: e.target.value
        });
    }

    render() {
    	const {marketfeedInstruments, selectedMarketfeedIndex} = this.props;
	let symbol = "";
	let bidPrice = null;
	let askPrice = null;
	let instrument = marketfeedInstruments[selectedMarketfeedIndex];

	if(instrument !== undefined) {
		symbol = instrument.symbol;
		bidPrice = instrument.bidPrice;
		askPrice = instrument.askPrice;
		if(this.state.symbol !== symbol) {
		    this.setState({symbol: symbol, price: instrument.price.toFixed(2)});
		}
	}
    return (
        <div id="ticket">
            <form>
		<p>
		    <label>Symbol:
			<input type="text" 
			       name="symbol" 
			       value={symbol}
			       onChange={this.symbolChange}/>
		    </label>
		</p>
		<p>
                    <label>Quantity:
			<input type="number" 
			       name="quantity" 
			       step="5" 
			       onChange={this.quantityChange} 
			       value={this.state.quantity} />
		    </label>
                </p>
		<p>
                    <label>Order Type:
		        <select id="orderType" name="order_type" onChange={this.orderTypeChange}>
			    <option value="Market">Market</option>
			    <option value="Limit">Limit</option>
			    <option value="Stop">Stop</option>
			    <option value="Stop Limit" >Stop Limit</option>
			</select>
		    </label> 
		</p>
		<p style={this.state.priceStyle}>
		    <label>Price:
		        <input type="number" 
			       name="price" 
			       step="0.1" 
			       onChange={this.priceChange} 
			       value={this.state.price} />
		    </label>
		</p>
		<p>
		    <label>ExpiryType:
		        <select name="expiryType" onChange={this.expiryTypeChange}>
		            <option value="Day">Day</option>
		            <option value="GTD">GTD</option>
		            <option value="GTC">GTC</option>
		            <option value="FOK">FOK</option>
		        </select>
                    </label> 
		</p>
                <p style={this.state.expiryDateStyle}>
		    <label>Expiry Date:
			<input type="date" 
				     name="expiryDate" 
				     onChange={this.expiryDateChange} 
				     value={this.state.expiryDate}/>
		    </label>
  		</p>
		<p>
                    <label>Note
			<textarea name="note" maxLength="100" onChange={this.noteChange} value={this.state.note}></textarea>
		    </label>
		</p>
		<div>
                    <div id="buySell">
		        <div id="sellButton" onClick={this.handleOnSubmit}>
			    <label>{bidPrice}</label>
			    <button id="buy">Sell</button>
                        </div>
                        <div id="buyButton" onClick={this.handleOnSubmit}>
			    <label>{askPrice}</label>
			    <button id="sell">Buy</button>
		        </div>
                    </div>
		</div>
	   </form>
	</div>
    )
    }
}
 
const mapStateToProps = state => ({ ...state });
// Map Redux actions to component props
const mapDispatchToProps = dispatch =>
    bindActionCreators({
       placeOrder,
       updateOrder
    }, dispatch);

// The HOC
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Ticket);


