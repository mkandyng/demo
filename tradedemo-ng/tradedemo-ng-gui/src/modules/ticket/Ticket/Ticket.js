import React, {useEffect} from "react";
import { getRandomInt, toggleOpacity, getDateString } from "../../../libs/utils";
import BuySellButton from "../../../components/BuySellButton";
import LabelInput from "../../../components/LabelInput";
import LabelTextArea from "../../../components/LabelTextArea";
import SelectionDropDown from "../../../components/SelectionDropDown";
import submitTicket from "../submitTicket";
import "./ticket.css";

/**
 * [Ticket Component to handle ticket view and interaction]
 */

export default function Ticket({ ticket,
                                 instrument,
                                 placeOrder,
                                 updateOrder,
                                 updateTicket }) {

    useEffect ( () => {
        const MAX_ORDER_COUNT = 10;
        if(instrument.symbol !== undefined && ticket.orderId < MAX_ORDER_COUNT) {
            let buySell = getRandomInt(0,1) === 0 ? "Buy":"Sell";
            if(submitTicket({ ticket: ticket,
                              instrument: instrument,
                              buySell: buySell,
                              confirmOrder: false,
                              placeOrder: placeOrder,
                              updateOrder: updateOrder })) {
                updateTicket({...ticket, orderId: ticket.orderId+1});
            }
        }
    }, [instrument, ticket, placeOrder, updateOrder, updateTicket]);

    const eventHandler = {
        symbolChange:     e => updateTicket({price: instrument.price.toFixed(2)}),
        expiryDateChange: e => updateTicket({expiryDate: e.target.value}),
        quantityChange:   e => updateTicket({price: e.target.value}),
        priceChange:      e => updateTicket({price: e.target.value}),
        noteChange:       e => updateTicket({note: e.target.value}),
        orderTypeChange:  e => updateTicket({
                  priceStyle: {opacity: toggleOpacity(e.target.value, "Market", true)},
                  orderType: e.target.value,
                  price: instrument.price.toFixed(2)}),
        expiryTypeChange: e => updateTicket({
                  expiryDateStyle: {opacity: toggleOpacity(e.target.value, "GTD", false)},
                  expiryType: e.target.value,
                  expiryDate: e.target.value === "GTD"?getDateString(new Date(), "dateOnly"):""}),
        handleOnSubmit:   e => {
                  e.preventDefault();
                  const buySell = e.target.innerText;
                  if(submitTicket({ ticket:ticket,
                                   instrument:instrument,
                                   buySell:buySell,
                                   confirmOrder: true,
                                   placeOrder: placeOrder,
                                   updateOrder: updateOrder})) {
                      updateTicket({orderId: ticket.orderId+1});
                  }
        }
    };

    return (
      <TicketView ticket={ticket}
                  instrument={instrument}
                  eventHandler={eventHandler} />
    )
}

export function TicketView({ticket, instrument, eventHandler}) {
    return (
        <div id="ticket">
            <form>
                <LabelInput label="Symbol"
                            name="symbol"
                            type="text"
                            value={instrument.symbol}
                            handleOnChange={eventHandler.symbolChange} />
                <LabelInput label="Quantity"
                            name="quantity"
                            type="number"
                            step={5}
                            value={ticket.quantity}
                            handleOnChange={eventHandler.quantityChange} />
                <SelectionDropDown id="orderType"
                                   label="Order Type"
                                   name="orderType"
                                   list={["Market","Limit","Stop","Stop Limit"]}
                                   handleOnChange={eventHandler.orderTypeChange} />
                <LabelInput style={ticket.priceStyle}
                            label="Price"
                            name="price"
                            type="number"
                            step={0.1}
                            value={ticket.price}
                            handleOnChange={eventHandler.priceChange} />
                <SelectionDropDown id="expiryType"
                                   label="Expiry Type"
                                   name="expiryType"
                                   list={["Day","GTD","GTC","FOK"]}
                                   handleOnChange={eventHandler.expiryTypeChange} />
                <LabelInput style={ticket.expiryDateStyle}
                            label="Expiry Date"
                            name="expiryDate"
                            type="date"
                            step={0.1}
                            value={ticket.expiryDate}
                            handleOnChange={eventHandler.expiryDateChange} />
                <LabelTextArea
                           label="Note"
                           name="note"
                           maxLength={100}
                           value={ticket.note}
                           handleOnChange={eventHandler.noteChange} />

                <div>
                    <div id="buySell">
                        <BuySellButton containerId="sellButton"
                                       handleOnClick={eventHandler.handleOnSubmit}
                                       label={instrument.bidPrice}
                                       id="sell"
                                       buttonName="Sell" />
                        <BuySellButton containerId="buyButton"
                                       handleOnClick={eventHandler.handleOnSubmit}
                                       label={instrument.askPrice}
                                       id="buy"
                                       buttonName="Buy" />
                    </div>
                </div>
            </form>
        </div>
    )
}
