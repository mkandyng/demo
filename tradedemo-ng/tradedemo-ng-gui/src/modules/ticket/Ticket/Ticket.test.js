import React from "react";
import toJson from "enzyme-to-json";
import {shallow} from "enzyme";
import Ticket, {TicketView} from "./Ticket";

/**
 * This is unit test of Orderbook component
 * It leverage snapshot to test the presentation
 */

describe("Ticket", () => {
  const ticket = {
    symbol: "symbol",
    quantity: 5,
    orderType: "Market",
    price: 0,
    expiryType: "Day",
    expiryDate: "2019-10-12",
    note: ""
  };
  const props = {
    ticket: ticket,
    instrument: ticket,
    placeOrder: jest.fn(),
    updateOrder: jest.fn(),
    updateTicket: jest.fn()
  };

  it("should render component comparing with previous snapshot shallow markup", () => {
    // Given

    // When
    const component = shallow(<Ticket {...props}/>);

    // Then
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render TicketView component comparing with previous snapshot shallow markup", () => {
    // Given
    const localProps = {
      ticket: ticket,
      instrument: ticket,
      eventHandler: {
        handleFormChange: jest.fn(),
        orderTypeChange: jest.fn(),
        expiryTypeChange: jest.fn(),
        handleOnSubmit: jest.fn()
      }
    };

    // When
    const component = shallow(<TicketView {...localProps}/>);

    // Then
    expect(toJson(component)).toMatchSnapshot();
  });

});
