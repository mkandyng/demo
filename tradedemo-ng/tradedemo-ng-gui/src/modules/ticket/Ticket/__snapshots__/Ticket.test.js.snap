// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket should render TicketView component comparing with previous snapshot shallow markup 1`] = `
<div
  id="ticket"
>
  <form>
    <LabelInput
      handleOnChange={[MockFunction]}
      label="Symbol"
      name="symbol"
      type="text"
      value="symbol"
    />
    <LabelInput
      handleOnChange={[MockFunction]}
      label="Quantity"
      name="quantity"
      step={5}
      type="number"
      value={5}
    />
    <SelectionDropDown
      handleOnChange={[MockFunction]}
      id="orderType"
      label="Order Type"
      list={
        Array [
          "Market",
          "Limit",
          "Stop",
          "Stop Limit",
        ]
      }
      name="orderType"
    />
    <LabelInput
      handleOnChange={[MockFunction]}
      label="Price"
      name="price"
      step={0.1}
      type="number"
      value={0}
    />
    <SelectionDropDown
      handleOnChange={[MockFunction]}
      id="expiryType"
      label="Expiry Type"
      list={
        Array [
          "Day",
          "GTD",
          "GTC",
          "FOK",
        ]
      }
      name="expiryType"
    />
    <LabelInput
      handleOnChange={[MockFunction]}
      label="Expiry Date"
      name="expiryDate"
      step={0.1}
      type="date"
      value="2019-10-12"
    />
    <LabelTextArea
      handleOnChange={[MockFunction]}
      label="Note"
      maxLength={100}
      name="note"
      value=""
    />
    <div>
      <div
        id="buySell"
      >
        <BuySellButton
          buttonName="Sell"
          containerId="sellButton"
          handleOnClick={[MockFunction]}
          id="sell"
        />
        <BuySellButton
          buttonName="Buy"
          containerId="buyButton"
          handleOnClick={[MockFunction]}
          id="buy"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`Ticket should render component comparing with previous snapshot shallow markup 1`] = `
<TicketView
  eventHandler={
    Object {
      "expiryDateChange": [Function],
      "expiryTypeChange": [Function],
      "handleOnSubmit": [Function],
      "noteChange": [Function],
      "orderTypeChange": [Function],
      "priceChange": [Function],
      "quantityChange": [Function],
      "symbolChange": [Function],
    }
  }
  instrument={
    Object {
      "expiryDate": "2019-10-12",
      "expiryType": "Day",
      "note": "",
      "orderType": "Market",
      "price": 0,
      "quantity": 5,
      "symbol": "symbol",
    }
  }
  ticket={
    Object {
      "expiryDate": "2019-10-12",
      "expiryType": "Day",
      "note": "",
      "orderType": "Market",
      "price": 0,
      "quantity": 5,
      "symbol": "symbol",
    }
  }
/>
`;
